### DISCUSSION TITLE
Better Optimizing My Setup

### DISCUSSION THREAD

**Discussion by teamcoltra** [pocketbase/pocketbase#4429] [Q&A]
Okay so based on #4426 being closed as "Will-Not-Improve" I would love to make sure that the setup that I have is sane:

I'm wanting to write about 20,000,000 records for books. 

For collections I have:

Authors, Books, Awards, Places, Characters, Series and also tables to link Awards and Places to an individual book. So for each time I want to add a new book I currently make the following API calls:

1. Write Place (x number of places).  If "name" bounces as not-unique then send another request to get ID of place, otherwise get the ID.
2. Write Character (x number of characters). If "name" bounces as not-unique then move on. (is unique against book_id, character_id)
3. Write Awards (x number of awards). If "name" bounces as not-unique then send another request to get ID of place, otherwise get the ID.
4. Write Series. If "name" bounces as not-unique then send another request to get ID of place, otherwise get the ID.
5. Write Author (x number of Authors). If "name" bounces as not-unique then send another request to get ID of place, otherwise get the ID.
6. Write Book with all the other details about the book. With a relational call to series.
7. Write book_author to link author and book
8. Write book_awards to link book and awards
9. write book_place to link book and place

All of these are currently done through http rest calls.

I can speed this up a bit by running the code from the server that PocketBase is on. I don't know how much it would speed things up but since my code is in Go I could probably build Pocketbase into framework mode and execute them within Pocketbase without any TCP (even internally) which would probably speed things up.

But am I missing anything? Currently I'm doing at minimum 9 calls one book takes a few seconds. Which means 20,000,000 will take...  about 2 years. 

I'm not here expecting free DBA advice, but just checking if there isn't something undocumented or something in the documents I've missed about maybe some batch calls or something that could help speed this up. 

I was thinking of just writing one big sqlite file and injecting it but I was reading in another post that messing with Sqlite without building it within Pocketbase would basically mean I wont be able to interact with the database within Pocketbase. 

Thank you so much for taking the time to bother reading this let alone any help I can get. 

Edit: and thanks for moving this into a discussion, sorry for wasting your resources on my inability to look at the options.

**Comment by ganigeorgiev** [comment]
> Currently I'm doing at minimum 9 calls one book takes a few seconds.

Could you show a code sample how you are sending the requests?
Are you sending them concurrently per book or await for each book each request in a loop?

> I'm wanting to write about 20,000,000 records for books.

In what format are those 20,000,000 records?

As an alternative of creating a pure SQL import script, if these records are in CSV or JSON you can register a custom console command with [Go](https://pocketbase.io/docs/go-console-commands/) or [JS](https://pocketbase.io/docs/js-console-commands/) and import them programmatically using the [Record operations](https://pocketbase.io/docs/go-records/) (it will be definitely faster and more efficient compared to sending individual HTTP requests).
The same apply if they are stored in some other database (at least when using Go) since nothing is stopping you to initialize a connection to the external database and fetch the result directly from there. I'm not sure if it will help but as an example you could check this small migration tool that I've written for my other project - https://github.com/presentator/v2tov3migrate.

**Reply by ganigeorgiev** [reply]
> I was thinking of just writing one big sqlite file and injecting it but I was reading in another post that messing with Sqlite without building it within Pocketbase would basically mean I wont be able to interact with the database within Pocketbase.

As long as you've created upfront the collections structure from the PocketBase Admin UI, you can safely import your raw SQL script into the corresponding SQLite tables (_each PocketBase collection has a table with the collection name_).

**Reply by teamcoltra** [reply]
> Could you show a code sample of how you are sending the requests

This is the main bit of processing

```
	seriesID, err := createOrFindEntity("series", map[string]interface{}{"name": book.Series.Name})
	if err != nil {
		return fmt.Errorf("error processing series: %w", err)
	}

	// Process authors
	var authorIDs []string
	for _, author := range book.Authors {
		authorData := map[string]interface{}{
			"name": author.Name,
			"role": author.Role,
			"bio":  author.Bio,
		}
		authorID, err := createOrFindEntity("authors", authorData)
		if err != nil {
			return fmt.Errorf("error processing author '%s': %w", author.Name, err)
		}
		authorIDs = append(authorIDs, authorID)
	}

	// Process awards
	var awardIDs []string
	awardNameToID := make(map[string]string)
	for _, award := range book.Awards {
		awardData := map[string]interface{}{
			"name": award.Name,
		}
		awardID, err := createOrFindEntity("awards", awardData)
		if err != nil {
			return fmt.Errorf("error processing award '%s': %w", award.Name, err)
		}
		awardIDs = append(awardIDs, awardID)
		awardNameToID[award.Name] = awardID // Store the mapping from award name to ID
	}

	// Process places
	var placeIDs []string
	for _, place := range book.Places {
		placeData := map[string]interface{}{
			"name": place.Name,
		}
		placeID, err := createOrFindEntity("places", placeData)
		if err != nil {
			return fmt.Errorf("error processing place '%s': %w", place.Name, err)
		}
		placeIDs = append(placeIDs, placeID)
	}

	// Create the book record
	bookData := map[string]interface{}{
		"title":                book.Title,
		"title_without_series": book.TitleWithoutSeries,
		"original_title":       book.OriginalTitle,
		"series_work_title":    book.Series.SeriesWorkTitle,
		"isbn":                 book.ISBN,
		"isbn13":               book.ISBN13,
		"asin":                 book.ASIN,
		"kindle_asin":          book.KindleASIN,
		"marketplace_id":       book.MarketplaceID,
		"language_code":        book.LanguageCode,
		"country_code":         book.CountryCode,
		"publication_day":      book.PublicationDay,
		"publication_month":    book.PublicationMonth,
		"publication_year":     book.PublicationYear,
		"is_cozy":              "FALSE",
		"is_rad":               "FALSE",
		"is_ebook":             book.IsEbook,
		"description":          book.Description,
		"publisher":            book.Publisher,
		"num_pages":            book.NumPages,
		"format":               book.Format,
		"language":             book.EditionLanguage,
		"series_id":            seriesID,
	}
	bookID, err := createOrFindEntity("books", bookData)
	if err != nil {
		return fmt.Errorf("error creating book '%s': %w", book.Title, err)
	}

	// Link authors to the book
	for _, authorID := range authorIDs {
		_, err := makePostRequest(fmt.Sprintf("%s/book_authors/records", baseURL), []byte(fmt.Sprintf(`{"book_id":"%s","author_id":"%s"}`, bookID, authorID)))
		if err != nil {
			return fmt.Errorf("error linking author '%s' to book '%s': %w", authorID, book.Title, err)
		}
	}

	// Link awards to the book
	for _, award := range book.Awards {
		awardID, exists := awardNameToID[award.Name]
		if !exists {
			return fmt.Errorf("awardID for '%s' does not exist", award.Name)
		}
		awardLinkData := fmt.Sprintf(`{"book_id":"%s","award_id":"%s","year":%d,"category":"%s"}`, bookID, awardID, award.Year, award.Category)
		_, err := makePostRequest(fmt.Sprintf("%s/book_awards/records", baseURL), []byte(awardLinkData))
		if err != nil {
			return fmt.Errorf("error linking award '%s' to book '%s': %w", award.Name, book.Title, err)
		}
	}

	// Link places to the book
	for _, placeID := range placeIDs {
		_, err := makePostRequest(fmt.Sprintf("%s/book_places/records", baseURL), []byte(fmt.Sprintf(`{"book_id":"%s","place_id":"%s"}`, bookID, placeID)))
		if err != nil {
			return fmt.Errorf("error linking place '%s' to book '%s': %w", placeID, book.Title, err)
		}
	}

	for _, character := range book.Characters {
		characterData := map[string]interface{}{
			"name":    character.Name,
			"book_id": bookID, // Ensures that each character is unique per book
		}
		_, err := createOrFindEntity("characters", characterData)
		if err != nil {
			return fmt.Errorf("error processing character '%s' for book '%s': %w", character.Name, book.Title, err)
		}
	}

```
This is the createorFindEntry function that I'm using for most of it:

```
// Function to create or find an entity and return its ID
	createOrFindEntity := func(collection string, data map[string]interface{}) (string, error) {
		fmt.Printf("Attempting to create/find entity in collection '%s' with data: %+v\n", collection, data)
		jsonData, err := json.Marshal(data)
		if err != nil {
			fmt.Printf("Error marshalling data: %v\n", err)
			return "", err
		}
		makeGetRequest := func(url string) (map[string]interface{}, error) {
			req, err := http.NewRequest("GET", url, nil)
			if err != nil {
				fmt.Printf("Error creating GET request: %v\n", err)
				return nil, err
			}
			req.Header.Set("Authorization", authHeader)
			req.Header.Set("Content-Type", "application/json")

			resp, err := client.Do(req)
			if err != nil {
				fmt.Printf("Error executing GET request: %v\n", err)
				return nil, err
			}
			defer resp.Body.Close()

			var result map[string]interface{}
			if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
				fmt.Printf("Error decoding GET response: %v\n", err)
				return nil, err
			}

			if resp.StatusCode != http.StatusOK {
				fmt.Printf("GET request failed with status code: %d\n", resp.StatusCode)
				return result, fmt.Errorf("request failed with status code: %d", resp.StatusCode)
			}

			fmt.Println("Successfully executed GET request.")
			return result, nil
		}

		result, err := makePostRequest(fmt.Sprintf("%s/%s/records", baseURL, collection), jsonData)
		if err != nil {
			// Modify the condition to check if the error string contains "400"
			if result != nil && strings.Contains(err.Error(), "400") {
				findURL := fmt.Sprintf("%s/%s/records?page=1&perPage=1&filter=name=\"%s\"&fields=id", baseURL, collection, url.QueryEscape(data["name"].(string)))
				findResult, findErr := makeGetRequest(findURL)
				if findErr != nil {
					fmt.Printf("Error during find operation: %v\n", findErr)
					return "", findErr // An error occurred during the find operation
				}
				fmt.Printf("Was there a result for the find: %v\n", findResult)
				items, ok := findResult["items"].([]interface{})
				if !ok || len(items) == 0 {
					fmt.Println("Entity not found after failed creation attempt due to 400 error")
					return "", fmt.Errorf("entity not found after failed creation attempt due to 400 error")
				}
				firstItem := items[0].(map[string]interface{})
				entityID, idOk := firstItem["id"].(string)
				if !idOk {
					fmt.Println("Could not extract ID from the found entity")
					return "", fmt.Errorf("could not extract ID from the found entity")
				}

				// Now you should return the correct ID
				return entityID, nil
			}
			return "", err
		}

		fmt.Printf("Successfully created entity with ID: %s\n", result["id"])
		return result["id"].(string), nil
	}
```
	They are basically just straight HTTP requests made by Go.
	
	
> In what format are those 20,000,000 records?

Currently they are all in xml files. I was going to convert them to json and just use them raw and have a simple database that called the raw files but then realized I should just do it right and put it into a database. 

I think what you suggested with the record operations makes sense, I could also do a custom "Framework" build of PB on my local PC (which is more powerful than my hosted server anyway) and build out the database using operations and then assuming I'm not missing anything just copy the files over to PB. It's fine if things get overwritten or whatever because I'm not using PB for anything other than this. 

It's still a big huge file transfer but probably better than the current setup.

**Reply by ganigeorgiev** [reply]
No matter whether you'll decide to use the Record operations or sending HTTP requests, I'd recommend to split your code and execute the different portions concurrently (eg. either per single book and/or per request operation basis).

Here is one pseudo-code example using [`errgroup.Group`](https://pkg.go.dev/golang.org/x/sync/errgroup) to resolve the `placeIDs` concurrently:

```go
var eg errgroup.Group

eg.SetLimit(1000)

// since we know upfront the number of places we can prealocate it
// and insert the id with the loop index to avoid using
// channel or mutex for the concurrent writes
var placeIDs = make([]string, len(book.Places))
for i, place := range book.Places {
    // shouldn't be needed with Go 1.22
    place := place
    i := i
    eg.Go(func() error {
        placeData := map[string]interface{}{
            "name": place.Name,
        }
        placeID, err := createOrFindEntity("places", placeData)
        if err != nil {
            return fmt.Errorf("error processing place %q: %w", place.Name, err)
        }
        placeIDs[i] = placeID
        return nil
    })
}
if err := eg.Wait(); err != nil {
    return err
}
```
